vendor/github.com/OpenBazaar/multiwallet/bitcoin/wallet.go

```diff
@@ -325,16 +325,20 @@ func (w *BitcoinWallet) GenerateMultisigScript(keys []hd.ExtendedKey, threshold
 	return w.generateMultisigScript(keys, threshold, timeout, timeoutKey)
 }
 
-func (w *BitcoinWallet) AddWatchedAddress(addr btc.Address) error {
-	script, err := w.AddressToScript(addr)
-	if err != nil {
-		return err
-	}
-	err = w.db.WatchedScripts().Put(script)
-	if err != nil {
-		return err
+func (w *BitcoinWallet) AddWatchedAddresses(addrs ...btc.Address) error {
+
+	for _, addr := range addrs {
+		script, err := w.AddressToScript(addr)
+		if err != nil {
+			return err
+		}
+		err = w.db.WatchedScripts().Put(script)
```

Comment:
This needs to be within a DB transaction so we don't get partial sets commited on error.

Reply:
Always got to add some design pattern we don't currently do at all don't you?

Reply:
Feel free to search the codebase for `db.Begin()` and say that again. ;)

Reply:
We do this for the wallet stuff?

Reply:
We do it at the put level but are you asking to make a PutScripts method in here that protects multiple simultaneously?

Reply:
```
func (w *WatchedScriptsDB) Put(scriptPubKey []byte) error {
	w.lock.Lock()
	defer w.lock.Unlock()
	tx, _ := w.db.Begin()
	stmt, err := tx.Prepare("insert or replace into watchedscripts(coin, scriptPubKey) values(?,?)")
	if err != nil {
		tx.Rollback()
		return err
	}
	defer stmt.Close()
	_, err = stmt.Exec(w.coinType.CurrencyCode(), hex.EncodeToString(scriptPubKey))
	if err != nil {
		tx.Rollback()
		return err
	}
	tx.Commit()
	return nil
}
```

Reply:
Yeah, it's in our Datastore which is exposed to the multiwallet. See:
```
repo/db/stxo.go
28:     tx, _ := s.db.Begin()

repo/db/txns.go
26:     tx, err := t.db.Begin()
138:    tx, err := t.db.Begin()

repo/db/watched_scripts.go
24:     tx, _ := w.db.Begin()

repo/db/utxos.go
28:     tx, _ := u.db.Begin()

repo/db/txmetadata.go
21:     tx, _ := t.db.Begin()
```

Reply:
But we already do it on the Put() for watchedscripts.

Reply:
This wasn't needed before because we weren't in a loop, it was always applying one addr and it was being checked for error at each step. Now we need to ensure we're not going to error halfway through our batch of addrs, otherwise we could error halfway through and our database will have a partial update.

Reply:
The transaction in WatchedScripts is a bit useless since it's already transactional (it only ever applies one record). Now we're making many calls to WatchScript each with a single addr. We could `Put(a)`, then `Put(b)` which errors, which cases `c, d, e, ...` to never get committed, and worse `a` would be added. The transaction needs to be removed from WatchedScripts and brought one level higher so the transaction can exist through all of the necessary Put calls.

Reply:
Yes that's what I was referring to. I was asking if you wanted to move it up a level. Ok.

Reply:
We decided this approach with PutAll() would be ok.